File: ED II/Material/Códigos/Árvores/percurso-arvores.c
#include <stdio.h>
#include <stdlib.h>

typedef struct No {
    int val;
    struct No* esq;
    struct No* dir;
} No;

No* novoNo(int val) {
    No* no = (No*)malloc(sizeof(No));
    no->val = val;
    no->esq = NULL;
    no->dir = NULL;
    return no;
}

void preOrdem(No* raiz) {
    if (raiz != NULL) {
        printf("%d ", raiz->val);
        preOrdem(raiz->esq);
        preOrdem(raiz->dir);
    }
}

void emOrdem(No* raiz) {
    if (raiz != NULL) {
        emOrdem(raiz->esq);
        printf("%d ", raiz->val);
        emOrdem(raiz->dir);
    }
}

void posOrdem(No* raiz) {
    if (raiz != NULL) {
        posOrdem(raiz->esq);
        posOrdem(raiz->dir);
        printf("%d ", raiz->val);
    }
}

int main() {
    
    No* raiz = novoNo(1);
    raiz->esq = novoNo(2);
    raiz->dir = novoNo(3);
    raiz->esq->esq = novoNo(4);
    raiz->esq->dir = novoNo(5);

    printf("Travessia pré-ordem: ");
    preOrdem(raiz);
    printf("\n");
 
    printf("Travessia em-ordem: ");
    emOrdem(raiz);
    printf("\n");
 
    printf("Travessia pós-ordem: ");
    posOrdem(raiz);
    printf("\n");
    return 0;
}

File: ED II/Material/Códigos/Árvores/percurso-arvores.py
class No:
    def __init__(self, val):
        self.val = val
        self.esq = None
        self.dir = None

def preOrdem(raiz):
    if raiz is not None:
        print(raiz.val, end=" ")
        preOrdem(raiz.esq)
        preOrdem(raiz.dir)

def emOrdem(raiz):
    if raiz is not None:
        emOrdem(raiz.esq)
        print(raiz.val, end=" ")
        emOrdem(raiz.dir)

def posOrdem(raiz):
    if raiz is not None:
        posOrdem(raiz.esq)
        posOrdem(raiz.dir)
        print(raiz.val, end=" ")

if __name__ == "__main__":

    raiz = No(1)
    raiz.esq = No(2)
    raiz.dir = No(3)
    raiz.esq.esq = No(4)
    raiz.esq.dir = No(5)

    print("Travessia Pré-Ordem: ", end="")
    preOrdem(raiz)
    print()

    print("Travessia Em-Ordem: ", end="")
    emOrdem(raiz)
    print()

    print("Travessia Pós-Ordem: ", end="")
    posOrdem(raiz)
    print()